tests/monster_vision.cpp

```diff
@@ -0,0 +1,48 @@
+#include "catch/catch.hpp"
+
+#include "game.h"
+#include "map.h"
+#include "map_helpers.h"
+#include "monster.h"
+#include "options_helpers.h"
+
+static monster &spawn_and_clear( const tripoint &pos, bool set_floor )
+{
+    if( set_floor ) {
+        g->m.set( pos, t_floor, f_null );
+    }
+    return spawn_test_monster( "mon_zombie", pos );
+}
+
+TEST_CASE( "monsters shouldn't see through floors" )
+{
+    clear_creatures();
+    monster &upper = spawn_and_clear( { 5, 5, 0 }, true );
+    monster &adjacent = spawn_and_clear( { 5, 6, 0 }, true );
+    monster &distant = spawn_and_clear( { 5, 3, 0 }, true );
+    monster &lower = spawn_and_clear( { 5, 5, -1 }, true );
+    monster &deep = spawn_and_clear( { 5, 5, -2 }, true );
+    monster &sky = spawn_and_clear( { 5, 5, 1 }, false );
+
+    // First check monsters whose vision should be blocked.
+    // One intervening floor between monsters.
+    CHECK( !upper.sees( lower ) );
+    CHECK( !lower.sees( upper ) );
+    // Two intervening floors between monsters.
+    CHECK( !upper.sees( deep ) );
+    CHECK( !deep.sees( upper ) );
+    // One intervening floor and an open space between monsters.
+    CHECK( !sky.sees( lower ) );
+    CHECK( !lower.sees( sky ) );
+
+    // Then cases where they should be able to see each other.
+    // No floor between monsters
+    CHECK( upper.sees( sky ) );
+    CHECK( sky.sees( upper ) );
+    // Adjacent monsters.
+    CHECK( upper.sees( adjacent ) );
+    CHECK( adjacent.sees( upper ) );
+    // distant monsters.
+    CHECK( upper.sees( distant ) );
+    CHECK( distant.sees( upper ) );
+}
```

Comment:
Hypothetically speaking, do you think it would be a good idea to also check that `lower` (or `deep`) also do not see `adjacent` nor `distant` to cover a possible bug where a light operation might not consider the floor on diagonal angles ?

Yes, it's redundant because of the check on `creature.cpp` requires both to be on the same Z-level, but with progress on Experimental Z levels I could see this in the future being changed to depend only on vision (not used to the codebase, so mostly guess work here) 🤔

Reply:
Yea, I have it open right now because of a config bug that was making it fail (idk, fields maybe?), so I'll add some cross checking.

Reply:
Well crap, now one more test is failing :/

Reply:
Thanks for poking me to check this, as a result I uncovered and fixed a *6 year old bug* in the foundational line drawing code.  There are probably dozens of latent bugs that this fixed.